pkg/server/jwk_resolver.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+
+	"github.com/abcxyz/pkg/cache"
+)
+
+type JWKResolver interface {
+	ResolveKeySet(ctx context.Context, oidcHeader string) (jwk.Set, error)
+}
+
+type OIDCResolver struct {
+	cache           *jwk.Cache
+	issuerToJwksURI *cache.Cache[string]
+}
+
+type OpenIDConfiguration struct {
+	// The only field we need from the config
+	JwksURI string `json:"jwks_uri"`
+}
+
+func NewOIDCResolver(ctx context.Context) *OIDCResolver {
+	return &OIDCResolver{
+		cache: jwk.NewCache(ctx),
+		// 24 hour cache timeout, jwks_uri should change very infrequently
+		issuerToJwksURI: cache.New[string](time.Hour * 24),
+	}
+}
+
+func (r *OIDCResolver) ResolveKeySet(ctx context.Context, oidcHeader string) (jwk.Set, error) {
+	issuer, err := r.extractIssuer(ctx, oidcHeader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract issuer from token: %w", err)
+	}
+
+	jwksURI, err := r.registerIssuer(ctx, issuer)
+	if err != nil {
+		return nil, fmt.Errorf("failed to register issuer %q: %w", issuer, err)
+	}
+
+	return jwk.NewCachedSet(r.cache, jwksURI), nil
+}
+
+func (r *OIDCResolver) extractIssuer(ctx context.Context, oidcHeader string) (string, error) {
+	// Parse without validation to extract issuer so we can use it to obtain the key set to use for validation
+	token, err := jwt.ParseString(oidcHeader, jwt.WithContext(ctx), jwt.WithVerify(false))
+	if err != nil {
+		return "", fmt.Errorf("failed to parse jwt header: %w", err)
+	}
+	return token.Issuer(), nil
+}
+
+func (r *OIDCResolver) registerIssuer(ctx context.Context, issuer string) (string, error) {
+	jwksURI, err := r.resolveJwksURI(ctx, issuer)
+	if err != nil {
+		return "", fmt.Errorf("failed to resolve JWKS URI for %q: %w", issuer, err)
+	}
+
+	if !r.cache.IsRegistered(jwksURI) {
+		err = r.cache.Register(jwksURI)
+		if err != nil {
+			return "", fmt.Errorf("failed to register JWKS URI %q to cache: %w", jwksURI, err)
+		}
+	}
+
+	return jwksURI, nil
+}
+
+func (r *OIDCResolver) resolveJwksURI(ctx context.Context, issuer string) (string, error) {
+	uri, cached := r.issuerToJwksURI.Lookup(issuer)
```

Comment:
This is a data race, and I'm also not sure it's necessary to cache this.

Reply:
Resolved by using `WriteThruLookup` instead